pkg/alert/alert.go

```diff
@@ -0,0 +1,427 @@
+// Copyright 2019 Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alert
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/gardener/test-infra/pkg/util/slack"
+	"github.com/go-logr/logr"
+	"github.com/olekukonko/tablewriter"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+)
+
+type Alert struct {
+	log logr.Logger
+	cfg Config
+	ctx context.Context
+}
+
+//TestDetails describes a test
+type TestDetails struct {
+	Name                string  //Name test name
+	Context             string  //Context is the concatenation of name and several test dimensions
+	FailedContinuously  bool    //FailedContinuously true if n recent test runs were failing in a row
+	LastFailedTimestamp string  //LastFailedTimestamp timestamp of last failed test execution
+	SuccessRate         float64 //SuccessRate of recent n days
+	Cloudprovider       string
+	OperatingSystem     string
+	Landscape           string
+	K8sVersion          string
+}
+
+type ElasticsearchConfig struct {
+	Authorization string //Authorization basic auth token in format "Basic R2fyZGVUZXI6bWR7Y1IxWkgycGpsNTdNNG1DbnQ="
+	Endpoint      string
+}
+
+//Config represents alerting configuration
+type Config struct {
+	EvalTimeDays                int //EvalTimeDays time range to consinder for evaluation in days (now - n days before)
+	SuccessRateThresholdPercent int //SuccessRateThresholdPercent if test success rate falls below threshold post an alert
+	ContinuousFailureThreshold  int //ContinuousFailureThreshold if test fails >=n times send alert
+	Elasticsearch               ElasticsearchConfig
+	Logger                      logr.Logger
+	Context                     context.Context
+	TestsToExclude              []string
+}
+
+func New(cfg Config) (*Alert, error) {
```

Comment:
```suggestion
func New(log logr.Logger, ctx context.Context, cfg Config) (*Alert, error) {
```

so you do not have duplicated refercnes in cfg and the alert struct itself